src/VisualStudio/Core/Def/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -492,8 +507,28 @@ private async Task EnsureWindowFrameInfoBarUpdatedAsync(CancellationToken cancel
             // Remove the current message so it can be replaced with the new one.
             _currentInfoBarMessage?.Remove();
 
-            _currentInfoBarMessage = await _infoBar.ShowInfoBarMessageAsync(
-                message, isCloseButtonVisible: false, imageMoniker).ConfigureAwait(true);
+            VisualStudioInfoBar.InfoBarMessage? infoBarMessage = null;
+            InfoBarUI[] infoBarItems = [new InfoBarUI(ServicesVSResources.Rerun_generator, InfoBarUI.UIKind.Button, () =>
+            {
+                _fileManager._threadingContext.ThrowIfNotOnUIThread();
+                Contract.ThrowIfNull(infoBarMessage);
+                infoBarMessage.Remove();
+
+                _currentInfoBarMessage = _fileManager._threadingContext.JoinableTaskFactory.Run(() =>
+                    _infoBar.ShowInfoBarMessageAsync(
+                        ServicesVSResources.Generator_running, isCloseButtonVisible: false, KnownMonikers.StatusInformation));
+
+                // Force regeneration here.  Nothing has actually changed, so the incremental generator architecture
+                // would normally just return the same values all over again.  By forcing things, we drop the
+                // generator driver, which will force new files to actually be created.
+                this.Workspace.EnqueueUpdateSourceGeneratorVersion(
+                    this._documentIdentity.DocumentId.ProjectId,
+                    forceRegeneration: true);
+            })];
+
+            infoBarMessage = await _infoBar.ShowInfoBarMessageAsync(
```

Comment:
>infoBarMessage [](http://example.com/codeflow?start=12&length=14)

Just wanted to make sure the local is really necessary. I'm not seeing the case where just using _currentInfoBarMessage directly doesn't work.

Reply:
teh reason for this is so we can call this above in the button callback:

```
                Contract.ThrowIfNull(infoBarMessage);
                infoBarMessage.Remove();
```

i like thsi style to make it clear that the button only refers to *this* instance of the message object, and couldn't accidentally/unintentionally refer/remove some other instance.

Reply:
i added docs to explain this.